internal/controller/pkg/revision/reconciler.go

```diff
@@ -615,40 +563,39 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 	pullSecretConfig, pullSecretFromConfig, err := r.config.PullSecretFor(ctx, pr.GetResolvedSource())
 	if err != nil {
 		err = errors.Wrap(err, errGetPullConfig)
-		status.MarkConditions(v1.Unhealthy().WithMessage(err.Error()))
+		status.MarkConditions(v1.RevisionUnhealthy().WithMessage(err.Error()))
 		_ = r.client.Status().Update(ctx, pr)
 
 		r.record.Event(pr, event.Warning(reasonImageConfig, err))
 
 		return reconcile.Result{}, err
 	}
 
-	var runtimeManifestBuilder ManifestBuilder
-	pwr, hasRuntime := pr.(v1.PackageRevisionWithRuntime)
-	if hasRuntime && r.runtimeHook != nil {
-		opts, err := r.runtimeManifestBuilderOptions(ctx, pwr)
-		if err != nil {
-			log.Debug(errManifestBuilderOptions, "error", err)
-
-			err = errors.Wrap(err, errManifestBuilderOptions)
-			status.MarkConditions(v1.Unhealthy().WithMessage(err.Error()))
-			_ = r.client.Status().Update(ctx, pr)
-
-			r.record.Event(pr, event.Warning(reasonSync, err))
-
-			return reconcile.Result{}, err
+	// Determine the desired pull secret config ref state
+	var psRef *v1.ImageConfigRef
+	if pullSecretConfig != "" {
+		psRef = &v1.ImageConfigRef{
+			Name:   pullSecretConfig,
+			Reason: v1.ImageConfigReasonSetPullSecret,
 		}
+	}
 
-		if pullSecretFromConfig != "" {
-			opts = append(opts, RuntimeManifestBuilderWithPullSecrets(pullSecretFromConfig))
+	// Check if the current applied image config ref for pull secret needs updating
+	curr := getCurrentImageConfigRef(pr, v1.ImageConfigReasonSetPullSecret)
+	if !imageConfigRefsEqual(curr, psRef) {
+		// Update the applied image config refs and persist immediately
+		pr.ClearAppliedImageConfigRef(v1.ImageConfigReasonSetPullSecret)
+		if psRef != nil {
+			log.Debug("Selected pull secret from image config store", "image", pr.GetResolvedSource(), "imageConfig", pullSecretConfig, "pullSecret", pullSecretFromConfig, "rewriteConfig", rewriteConfigName)
+			r.record.Event(pr, event.Normal(reasonImageConfig, fmt.Sprintf("Selected pullSecret %q from ImageConfig %q for registry authentication", pullSecretFromConfig, pullSecretConfig)))
+			pr.SetAppliedImageConfigRefs(*psRef)
 		}
-
-		runtimeManifestBuilder = NewRuntimeManifestBuilder(pwr, r.namespace, opts...)
+		return reconcile.Result{Requeue: true}, errors.Wrap(r.client.Status().Update(ctx, pr), errUpdateStatus)
```

Comment:
We need to ensure that applied image config for pull secret is persisted to the status since runtime reconciler relies on this instead of resolving in both places. 

Previously, it was not persisted since there was a client.Update before a status update.